File: python/src/bursar/credits/types/grants.py
"""Grant-program types — mirrors JS SDK's ``credits/types/grants.ts``."""

from __future__ import annotations

from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, model_validator

from bursar.credits.types.account import CreditMetadata

GrantProgramTrigger = Literal[
    "account_created",
    "referral_completed",
    "promo_code_redeemed",
    "manual",
]


class ExecuteGrantProgramRequest(BaseModel):
    """One application event that may award one or more catalog grants."""

    trigger: GrantProgramTrigger
    program_key: str
    subject_id: str
    event_key: str
    referrer_subject_id: str | None = None
    region: str | None = None
    metadata: CreditMetadata | None = None


class GrantProgramAwardResult(BaseModel):
    """One award row produced by a grant-program execution."""

    model_config = ConfigDict(extra="forbid")

    grant_event_id: str | None
    grant_award_id: str | None
    recipient_subject_id: str | None
    ledger_entry_id: str | None
    amount: Decimal | None
    replayed: bool
    error: str | None

    @model_validator(mode="after")
    def validate_outcome(self) -> GrantProgramAwardResult:
        award = (
            self.grant_event_id,
            self.grant_award_id,
            self.recipient_subject_id,
            self.ledger_entry_id,
            self.amount,
        )
        if self.error is None and any(value is None for value in award):
            raise ValueError("successful grant awards require committed fields")
        if self.error is not None and (any(value is not None for value in award) or self.replayed):
            raise ValueError("failed grant awards cannot expose committed fields")
        return self

File: javascript/src/credits/types/grants.ts
import type { Decimal } from "decimal.js";
import type { CreditMetadata } from "./account.js";

export type GrantProgramTrigger =
  | "account_created"
  | "referral_completed"
  | "promo_code_redeemed"
  | "manual";

/** One application event that may award one or more catalog grants. */
export interface ExecuteGrantProgramRequest {
  trigger: GrantProgramTrigger;
  programKey: string;
  subjectId: string;
  eventKey: string;
  referrerSubjectId?: string | null;
  region?: string | null;
  metadata?: CreditMetadata | null;
}

interface GrantProgramAwardBase {
  replayed: boolean;
}

/** One committed award produced by a grant-program execution. */
export interface GrantProgramAwardSuccess extends GrantProgramAwardBase {
  grantEventId: string;
  grantAwardId: string;
  recipientSubjectId: string;
  ledgerEntryId: string;
  amount: Decimal;
  error: null;
}

/** A rejected grant-program execution. No financial identifiers are fabricated. */
export interface GrantProgramAwardFailure extends GrantProgramAwardBase {
  grantEventId: null;
  grantAwardId: null;
  recipientSubjectId: null;
  ledgerEntryId: null;
  amount: null;
  replayed: false;
  error: string;
}

export type GrantProgramAwardResult = GrantProgramAwardSuccess | GrantProgramAwardFailure;
